src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1055 @@
+            let old_ratio = self.discard_ratio.load(Ordering::Relaxed);
+
+            // Because pending compaction bytes changes up and down, so using
+            // EMA(Exponential Moving Average) to smooth it.
+            (if old_ratio != 0 {
+                EMA_FACTOR * (old_ratio as f64 / RATIO_SCALE_FACTOR)
+                    + (1.0 - EMA_FACTOR) * new_ratio
+            } else {
+                if new_ratio > 0.01 { 0.01 } else { new_ratio } // cap the initial discard ratio
+            } * RATIO_SCALE_FACTOR) as u64
+        };
+        SCHED_DISCARD_RATIO_GAUGE.set(ratio as i64);
+        self.discard_ratio.store(ratio, Ordering::Relaxed);
+    }
+
+    fn adjust_memtables(&mut self, cf: &str) {
+        let num_memtables = self
+            .engine
+            .get_cf_num_immutable_mem_table(cf)
+            .unwrap_or(None)
+            .unwrap_or(0);
+        let checker = self.cf_checkers.get_mut(cf).unwrap();
+        SCHED_MEMTABLE_GAUGE
+            .with_label_values(&[cf])
+            .set(num_memtables as i64);
+        let prev = checker.last_num_memtables.get_recent();
+        checker.last_num_memtables.observe(num_memtables);
+
+        // do special check on start, see the comment of the variable definition for detail.
+        if checker.on_start_memtable {
+            if num_memtables < self.memtables_threshold
+                || checker.last_num_memtables.trend() == Trend::Increasing
+            {
+                // the write is accumulating, still need to throttle
+                checker.on_start_memtable = false;
+            } else {
+                // still on start, should not throttle now
+                return;
+            }
+        }
+
+        for c in self.cf_checkers.values() {
+            if num_memtables < c.last_num_memtables.get_recent() {
+                return;
+            }
+        }
+
+        let checker = self.cf_checkers.get_mut(cf).unwrap();
+        let is_throttled = self.limiter.speed_limit() != INFINITY;
+        let should_throttle =
+            checker.last_num_memtables.get_avg() > self.memtables_threshold as f64;
+        let throttle = if !is_throttled {
+            if should_throttle {
+                SCHED_THROTTLE_ACTION_COUNTER
+                    .with_label_values(&[cf, "memtable_init"])
+                    .inc();
+                checker.init_speed = true;
+                let x = self.write_flow_recorder.get_percentile_90();
+                if x == 0 { INFINITY } else { x as f64 }
+            } else {
+                INFINITY
+            }
+        } else if !should_throttle
+            || checker.last_num_memtables.get_recent() < self.memtables_threshold
+        {
+            // should not throttle memtable
+            checker.memtable_debt = 0.0;
+            if checker.init_speed {
+                INFINITY
+            } else {
+                self.limiter.speed_limit() + checker.memtable_debt
```

Comment:
This seems wrong, where is the 1MB factor? Looks like memtable throttle is not well tested.

Reply:
yes... hard to see memtable accumulates but L0 doesn't.

Reply:
Maybe use OS injection tool to stop disk writes? @5kbpers has done that before.

Reply:
This seems like reproducible: #9649. I'm thinking, if memtable stall is rare, how about using a super simple algorithm? e.g. discard all requests when memtables breach threshold.

Reply:
I think it's simple enough which closes to rocksdb impl. Will try to reproduce it later, if it doesn't fit well, change it then